compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNotImplementedOverrideChecker.kt

```diff
@@ -46,74 +48,15 @@ object FirNotImplementedOverrideChecker : FirClassChecker() {
-            if (containingClass is FirRegularClass && containingClass.isExpect) return false
-            return true
-        }
-
-        for (name in classScope.getCallableNames()) {
-            classScope.processFunctionsByName(name) { namedFunctionSymbol ->
-                val simpleFunction = namedFunctionSymbol.fir
-                if (namedFunctionSymbol is FirIntersectionOverrideFunctionSymbol) {
-                    if (namedFunctionSymbol.intersections.count {
-                            (it.fir as FirCallableMemberDeclaration).modality != Modality.ABSTRACT
-                        } > 1 && simpleFunction.getContainingClass(context) === declaration
-                    ) {
-                        notImplementedIntersectionSymbols += namedFunctionSymbol
-                        return@processFunctionsByName
-                    }
-                }
-                if (!simpleFunction.shouldBeImplemented()) return@processFunctionsByName
-                if (declaration is FirRegularClass && declaration.isData && simpleFunction.matchesDataClassSyntheticMemberSignatures) {
-                    return@processFunctionsByName
-                }
 
-                // TODO: suspend function overridden by a Java class in the middle is not properly regarded as an override
-                if (simpleFunction.isSuspend) return@processFunctionsByName
-                if (simpleFunction.isInvisible()) {
-                    invisibleSymbols += namedFunctionSymbol
-                } else {
-                    notImplementedSymbols += namedFunctionSymbol
-                }
-            }
-            classScope.processPropertiesByName(name) { propertySymbol ->
-                val property = propertySymbol.fir as? FirProperty ?: return@processPropertiesByName
-                if (propertySymbol is FirIntersectionOverridePropertySymbol) {
-                    if (propertySymbol.intersections.count {
-                            (it.fir as FirCallableMemberDeclaration).modality != Modality.ABSTRACT
-                        } > 1 && property.getContainingClass(context) === declaration
-                    ) {
-                        notImplementedIntersectionSymbols += propertySymbol
-                        return@processPropertiesByName
-                    }
-                }
-                if (!property.shouldBeImplemented()) return@processPropertiesByName
-
-                if (property.isInvisible()) {
-                    invisibleSymbols += propertySymbol
-                } else {
-                    notImplementedSymbols += propertySymbol
-                }
+        collectSymbols(declaration, context.sessionHolder) { symbol, isNotImplemented, isAmbiguouslyImplemented, isVisible, _ ->
+            when {
+                symbol is FirIntersectionCallableSymbol && isAmbiguouslyImplemented -> notImplementedIntersectionSymbols.add(symbol)
```

Comment:
Nit: after looking at the logic below, I wonder `symbol is FirIntersectionCallableSymbol` is not that necessary? If you split `shouldBeImplemented` back, I think something `isAmbiguous` here implies it `is FirIntersectionCallableSymbol`.

Reply:
Nice catch! Yep it's not needed.